File: backend/src/main/java/br/com/calciolari/datahub/catalog/application/ProductQueryService.java
package br.com.calciolari.datahub.catalog.application;

import java.util.List;
import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import br.com.calciolari.datahub.catalog.api.ProductDtos.ProductDetail;
import br.com.calciolari.datahub.catalog.api.ProductDtos.ProductSummary;
import br.com.calciolari.datahub.catalog.infrastructure.persistence.ProductEntity;
import br.com.calciolari.datahub.catalog.infrastructure.persistence.ProductRepository;
import br.com.calciolari.datahub.shared.api.PageResponse;

@Service
public class ProductQueryService {

	private final ProductRepository productRepository;

	public ProductQueryService(ProductRepository productRepository) {
		this.productRepository = productRepository;
	}

	public PageResponse<ProductSummary> list(String query, int page, int size) {
		Page<ProductEntity> result = productRepository.searchPublished(
				blankToNull(query), PageRequest.of(page, size, org.springframework.data.domain.Sort.by("name")));
		List<ProductSummary> content = result.getContent().stream()
				.map(p -> new ProductSummary(p.getId(), p.getExternalSource(), p.getExternalId(), p.getName(), null))
				.toList();
		return PageResponse.of(content, page, size, result.getTotalElements());
	}

	public ProductDetail get(UUID id) {
		ProductEntity product = productRepository.findPublishedById(id)
				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Product not found"));
		return new ProductDetail(
				product.getId(),
				product.getExternalSource(),
				product.getExternalId(),
				product.getName(),
				null,
				product.getFirstSeenParseAttemptId());
	}

	private static String blankToNull(String value) {
		return value == null || value.isBlank() ? null : value.trim();
	}
}

File: backend/src/Calciolari.DataHub/Catalog/Application/ProductQueryService.cs
using Calciolari.DataHub.Persistence;
using Calciolari.DataHub.Shared.Api;
using Microsoft.EntityFrameworkCore;

namespace Calciolari.DataHub.Catalog.Application;

public sealed class ProductQueryService
{
    private readonly DataHubDbContext _db;

    public ProductQueryService(DataHubDbContext db)
    {
        _db = db;
    }

    public PageResponse<ProductSummary> List(string? query, int page, int size)
    {
        var q = BlankToNull(query);
        var published = PublishedProducts();
        if (q is not null)
        {
            var like = "%" + q + "%";
            published = published.Where(p =>
                EF.Functions.ILike(p.Name, like) || EF.Functions.ILike(p.ExternalId, like));
        }

        var ordered = published.OrderBy(p => p.Name);
        var total = ordered.Count();
        var content = ordered.Skip(page * size).Take(size)
            .Select(p => new ProductSummary(p.Id, p.ExternalSource, p.ExternalId, p.Name, null))
            .ToList();
        return PageResponse<ProductSummary>.Of(content, page, size, total);
    }

    public ProductDetail Get(Guid id)
    {
        var product = PublishedProducts().SingleOrDefault(p => p.Id == id)
                      ?? throw new ApiException(StatusCodes.Status404NotFound, "Product not found");
        return new ProductDetail(
            product.Id,
            product.ExternalSource,
            product.ExternalId,
            product.Name,
            null,
            product.FirstSeenParseAttemptId);
    }

    private IQueryable<Persistence.Entities.ProductEntity> PublishedProducts() =>
        _db.Products.AsNoTracking().Where(p =>
            _db.SaleItems.Any(si =>
                si.ProductId == p.Id
                && _db.ArtifactPublications.Any(ap => ap.ActiveParseAttemptId == si.ParseAttemptId)));

    internal static string? BlankToNull(string? value) =>
        string.IsNullOrWhiteSpace(value) ? null : value.Trim();
}

public sealed record ProductSummary(
    Guid Id,
    string ExternalSource,
    string ExternalId,
    string Name,
    string? Unit);

public sealed record ProductDetail(
    Guid Id,
    string ExternalSource,
    string ExternalId,
    string Name,
    string? Unit,
    Guid FirstSeenParseAttemptId);
